knimin/lib/data_access.py

```diff
@@ -2443,35 +2446,38 @@ def _study_is_unique(self, qiita_study_id=None, title=None, skip_id=None):
         ValueError
             If neither value is given, or either value already exists
         """
-        tags = ['Qiita study ID', 'Title']
-        cols = [tag.lower().replace(' ', '_') for tag in tags]
-        vals = [vars()[col] for col in cols]
-        if all(val is None for val in vals):
-            raise ValueError('Either %s or %s must be specified.'
-                             % (tags[0], tags[1]))
-        dups = [0 for tag in tags]
-        for i in range(len(vals)):
-            if tags[i]:
-                sql = """SELECT study_id
-                         FROM pm.study
-                         WHERE {} = %s
-                         AND study_id IS DISTINCT FROM %s""".format(cols[i])
-                sql_args = [vals[i], skip_id]
-                res = self._con.execute_fetchone(sql, sql_args)
-                if res:
-                    dups[i] = res[0]
-        if dups[0] and len(set(dups)) <= 1:
-            raise ValueError('%s %s and %s %s conflict with exisiting study '
-                             '%s.' % (tags[0], repr(vals[0]), tags[1],
-                                      repr(vals[1]), dups[0]))
-        else:
-            errs = []
-            for i in range(len(tags)):
-                if dups[i]:
-                    errs.append('%s %s conflicts with exisiting study %s.'
-                                % (tags[i], repr(vals[i]), dups[i]))
-            if errs:
-                raise ValueError(' '.join(errs))
+        with TRN:
```

Comment:
This entire function can be substituted by:
```python
def _study_is_unique(self, qiita_study_id=None, title=None, skip_id=None):
    """DOCS..."""
    if qiita_study_id is None and title is None:
        raise ValueError('Either qiita_study_id or title must '
                         'be specified')

    with TRN:
        cols = ['qiita_study_id', 'title']
        vals = [qiita_study_id, title]
        sql = """SELECT study_id
                 FROM pm.study
                 WHERE {} = %s AND study_id IS DISTINCT FROM %s"""

        errors = []
        err_msg = "Study with %s '%s' already exists: %s"
        for col, val in zip(cols, vals):
            TRN.add(sql.format(col), [val, skip_id])
            res = TRN.execute_fetchflatten()
            if res:
                errors.append(err_msg % (col, val, res[0]))

        if errors:
            raise ValueError('\n'.join(errors))
```

Which in my opinion is more readable and easy to maintain. Also, need to update the documentation to reflect that there is another case in which ValueError can be raised